test/contrib/test_ssltransport.py

```diff
@@ -0,0 +1,393 @@
+        # Proxy server will handle the first TLS connection and create a
+        # connection to the destination server.
+        cls.proxy_server = SocketProxyDummyServer(cls.host, cls.port)
+        cls.proxy_server.start_proxy_handler()
+
+    @classmethod
+    def teardown_class(cls):
+        if hasattr(cls, "proxy_server"):
+            cls.proxy_server.teardown_class()
+        super(TlsInTlsTestCase, cls).teardown_class()
+
+    @classmethod
+    def start_destination_server(cls):
+        """
+        Socket handler for the destination_server. Terminates the second TLS
+        layer and send a basic HTTP response.
+        """
+
+        def socket_handler(listener):
+            sock = listener.accept()[0]
+            with cls.server_context.wrap_socket(sock, server_side=True) as ssock:
+                request = consume_socket(ssock)
+                assert request is not None
+                assert "www.testing.com" in request.decode("utf-8")
+
+                response = b"HTTP/1.1 200 OK\r\n" b"Content-Length: 0\r\n" b"\r\n"
+                ssock.send(response)
+                ssock.close()
+
+        cls._start_server(socket_handler)
+
+    def test_tls_in_tls_tunnel(self):
+        """
+        Basic communication over the TLS in TLS tunnel.
+        """
+        self.start_destination_server()
+        self.start_proxy_server()
+
+        sock = socket.create_connection(
+            (self.proxy_server.host, self.proxy_server.port)
+        )
+        with self.client_context.wrap_socket(
+            sock, server_hostname="localhost"
+        ) as proxy_sock:
+            with SSLTransport(
+                proxy_sock, self.client_context, server_hostname="localhost"
+            ) as destination_sock:
+                assert destination_sock.version() is not None
+                destination_sock.send(
+                    b"GET http://www.testing.com/ HTTP/1.1\r\n"
+                    b"Host: www.testing.com\r\n"
+                    b"User-Agent: awesome-test\r\n"
+                    b"\r\n"
+                )
+                response = consume_socket(destination_sock)
+                assert response is not None
+                assert "200" in response.decode("utf-8")
+
+    def test_wrong_sni_hint(self):
+        """
+        Provides a wrong sni hint to validate an exception is thrown.
+        """
+        self.start_destination_server()
+        self.start_proxy_server()
+
+        sock = socket.create_connection(
+            (self.proxy_server.host, self.proxy_server.port)
+        )
+        with self.client_context.wrap_socket(
+            sock, server_hostname="localhost"
+        ) as proxy_sock:
+            with pytest.raises(Exception) as e:
+                SSLTransport(
+                    proxy_sock, self.client_context, server_hostname="veryverywrong"
+                )
+            # Accommodate different python3 versions
+            assert e.type in [SSLTransportError, ssl.CertificateError]
```

Comment:
Why are we receiving two different error types here? Can we change this if we end up removing `SSLTransportError`?

Reply:
Prior to versions 3.6, SSL.CertificateError was actually a child of ValueError instead of SSLError. After 3.7 they changed it to be SSLError thankfully.

Reply:
Huh, I should have known that oddity :) Could this assert be `assert e.type == ssl.CertificateError`?